### ISSUE TITLE
RLMApp login marked with NS_REFINED_FOR_SWIFT

### ISSUE THREAD

**Issue by levelchanger** [realm/realm-cocoa#6857]
!!! MANDATORY TO FILL OUT !!!

## Goals

Starting with 10.0.0-rc.1, the login method of RLMApp has been marked with NS_REFINED_FOR_SWIFT and there is no Swiftified way to call the function. This change was not documented in the change log and it appears to affect more than just this method. Why was this done?

## Expected Results

Be able to call methods that are marked as NS_REFINED_FOR_SWIFT in Swift. I use to call the login function like so:

realmApp!.login(withCredential: credentials, completion: { (syncUser, error) in
})

## Actual Results

Now I have to use underscores to call the login method:

realmApp!.__login(withCredential: credentials, completion: { (syncUser, error) in
})

## Steps for others to Reproduce

See above.

## Code Sample

See above.

## Version of Realm and Tooling

Realm framework version: 10.0.0-rc.1

Realm Object Server version: NA

Xcode version: 12.0.1

iOS/OSX version: 10.15.7

Dependency manager + version: ?

**Comment by jsflax** [comment]
I can confirm that this is a bug. We'll have this fixed before final release (later this week).

**Comment by leemaguire** [comment]
Hi @levelchanger ,

The change description for this actually ended up under the v5.2.0 release notes, as seen here https://github.com/realm/realm-cocoa/blob/v10/CHANGELOG.md#520-release-notes-2020-06-30

The `App.login(withCredential:completion:)` method has changed to `App.login(credentials:completion:)`

**Comment by levelchanger** [comment]
This was not fixed in 10.0.0.rc-2. There are many methods marked as NS_REFINED_FOR_SWIFT for no apparent reason. I have a mixed Objective C/Swift project so I use the Objective C SDK. I've had to write wrappers in Objective C around the methods marked with NS_REFINED_FOR_SWIFT so I can call them from my Swift code. Why, oh why, are methods marked with NS_REFINED_FOR_SWIFT?

**Comment by ericjordanmossman** [comment]
> There are many methods marked as NS_REFINED_FOR_SWIFT for no apparent reason.

The reason is to steer RealmSwift developers away from calling RLM* versions when it could cause issues. As you pointed out, the mixed objc/swift is caught in the middle.

I've got a proposed solution [here](https://github.com/realm/realm-cocoa/pull/6886).
Because the vast majority of the Swift API are typealias of the ObjC version, you're still able to use Swift API for most operations. You'd continue to use RLMRealm, RLMConfiguration, and object models still inherit from RLMObject. But you'd use the RealmSwift for everything else -- `App`, `User`, `Credentials`, etc. Anything not covered by a typealias is then handled in ObjectiveCSupport.swift and RLMSupport.swift (which would be in the RealmSwift target and not manually downloaded) .

I'm curious if you (@levelchanger or @andreasley ) have any thoughts on this approach. The first thing is that you'd now have to include RealmSwift in a mixed objc/swift project, but the RealmSwift footprint is significantly smaller than `Realm`, and the benefits are pretty worth it.

**Comment by ericjordanmossman** [comment]
So for this particular issue the solution would be something like:
```
app.login(credentials) {result in
   // do something with user or error
}
```
^ RealmSwift.App and RealmSwift.Credentials instead of RLMApp and RLMCredentials. You'd get back a RealmSwift.User.
And later:
```
let config = user.configuration(partitionValue: appId) // A RealmSwift.User creates a Realm.Configuration
let rlmConfig = ObjectiveCSupport.convert(object: config) // Convert to RLMConfiguration
let rlmRealm = try! RLMRealm(configuration: rlmConfig)
```

**Comment by levelchanger** [comment]
I’m never thrilled to include another SDK but do I really have a choice at this point? 

This issue aside, it would be helpful to have some documentation for developers with mixed objc/swift projects. Right now, there isn’t anything.

**Comment by ianpward** [comment]
thank you for your comments @levelchanger - to give a little context on why more and more of our APIs are moving to favoring Swift, it it because approximately a year ago - 75% of our users were using Obj-C, now, only one year later this number is flipped with 75% using Swift and the remaining 25% using Obj-C. This aligns with the sentiment in the wider iOS community and we predict that a year from now very few users will be using Objective-C. 

While I realize that this is probably of little comfort to you, I share it in the vein of full transparency and while I don't see us writing a document on mixed objc/swift projects - I can say that we are happy to answer any questions you might have and encourage you to migrate to leveraging Swift fully in your application.

**Comment by levelchanger** [comment]
For some perspective, we started our app in 2011. Objective C was the only game in town for iOS and we built up a substantial code base in the language. Since Swift came out, we have used it for all new code development. The eventual goal is to move the whole project to Swift, but that is a big undertaking. Right now we are migrating from Core Data to Realm. We have 22 classes that descend from RLMObject. Most of those classes have methods that operate on the data. Could I switch to using the Swift Realm SDK and have a base class for each of our 22 classes that descends from Object and then have an Objective C subclass for each with the methods I already have? For example, say I have the following in Objective C:

@interface TDContact : RLMObject

@property RLMObjectId * _id;
@property NSString *_partition;
@property NSString * lastName;
@property NSString * firstName;

- (void)doSomethingWithThisData;
- (void)doSomethingElseWithThisData;

@end

Instead can I have a Swift base class with just the data?

class Contact: Object {
    @objc dynamic var _id: String = ""
    @objc dynamic var _partition: String = ""
    @objc dynamic var firstName: String = ""
    @objc dynamic var lastName: String = ""
    override static func primaryKey() -> String? {
        return "_id"
    }
}

And then have a Objective C subclass with the methods I have already written in Objective C, of which there are many.

@interface ObjCContact: Contact
- (void)doSomethingWithThisData;
- (void)doSomethingElseWithThisData;
@end

Would this work? 

Nina

**Comment by levelchanger** [comment]
Instead of subclassing I meant to say make a category or extension in Objective C of the class Contact that is in Swift.

Given that I have a mixed project do I have to have all of my code in Swift to use the Realm Swift SDK instead of the Objective C SDK?

**Comment by levelchanger** [comment]
Hi Ian,

Can I still have a mixed ObjC/Swift project and use the Realm Swift SDK?  If I convert our data classes from ObjC to Swift and then use them in ObjC view controllers, will that work? We haven’t released the app yet, so I don’t have an existing customer database to worry about. What is the catch with using the Realm Swift SDK with a mixed project?

Nina

**Comment by ianpward** [comment]
The RealmSwift And RealmObjectiveC APIs are not interoperable. So any Realm APIs need to be in a consistent language across the whole app. You can have code in another language used in the same app as long as it is not touching the Realm APIs

**Comment by levelchanger** [comment]
Is it OK to use an unmanaged Swift Realm object in ObjC code but do saving in Realm in Swift code?

**Comment by tgoyne** [comment]
https://github.com/realm/realm-cocoa/pull/6886 makes it so that in a mixed project you define your models as RLMObject subclasses and use the obj-c API for interacting with data regardless of language, but for everything sync related you use the language specific API. In other words, you can't use Object/Realm/List/Results/LinkingObjects from RealmSwift and have to use the RLM-prefixed versions instead, but can use things like App and User (although you can call them RLMApp and RLMUser if you want). The end result is very similar to how things worked with the obj-c api plus RLMSupport.swift, but with the Swift extensions for things marked NS_REFINED_FOR_SWIFT provided by the RealmSwift framework rather than a file you add to your project. The rationale for this change is just that RLMSupport.swift for v10 would end up being the bulk of RealmSwift anyway, so it no longer makes much sense for it to be a separate thing.

**Comment by ericjordanmossman** [comment]
#6886 released in [10.2.0](https://github.com/realm/realm-cocoa/releases/tag/v10.2.0)
